File: 03.CPP/0904.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

class Solution {
public:
    int totalFruit(std::vector<int>& fruits) {
        int n = fruits.size();

        int left = 0;
        int maxLen = 0;
        std::unordered_map<int, int>  fruitMap;

        for (int right = 0; right < n; right++) {
            int current = fruits[right];

            fruitMap[current]++;

            while (fruitMap.size() > 2) {
                fruitMap[fruits[left]]--;

                if (fruitMap[fruits[left]] == 0) {
                    fruitMap.erase(fruits[left]);
                }

                left++;
            }

            int currLen = right - left + 1;
            if (currLen > maxLen) {
                maxLen = currLen;
            }
        }

        return maxLen;
    }
};

int main() {
    Solution sol;
    
    std::vector<int> fruits = {1, 2, 3, 2, 2};
    int result = sol.totalFruit(fruits);

    std::cout << result << "\n";
    return 0;
}

File: 01.Python/0904.py
from typing import List

class Solution:
    def totalFruits(self, fruits: List[int]) -> int:
        n = len(fruits)

        fruitMap = {}
        left = 0
        maxLen = 0

        for right in range(n):
            current = fruits[right]

            if current in fruitMap:
                fruitMap[current] += 1
            else:
                fruitMap[current] = 1

            while len(fruitMap) >  2:
                fruitMap[fruits[left]] -= 1

                if fruitMap[fruits[left]] == 0:
                    del fruitMap[fruits[left]]
                left += 1

            currLen = right - left + 1
            if currLen > maxLen:
                maxLen = currLen
        
        return maxLen

sol = Solution()

fruits = [1, 2, 3, 2, 2]
print(sol.totalFruits(fruits))